### PR TITLE
Products APIs

### LINKED ISSUE #12: Products front stage API

### PR DESCRIPTION
**Pull request by asdfg429**
Products APIs Created

### CODE REVIEW

File: src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public Products getProductById(Long productId) {
+        Optional<Products> optionalUser =  productsRepository.findById(productId);
+        return optionalUser.get();
+    }
+
+    @Override
+    public Iterable<Products> getAllProducts() {
+        return productsRepository.findAll();
```
Comment (Vince-Ku) [inline src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java:35]: What if an error occurs in this case?

File: src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        TypedQuery<Products> query = entityManager.createQuery(jpql.toString(), Products.class);
+        query.setParameter("names", name);
+        query.setParameter("sortBy", sortBy);
+
+        return query.getResultList();*/
+        // TODO Auto-generated method stub
+        //productsRepository.getProductsByName(name, sortBy, orderBy);
+        //throw new UnsupportedOperationException("Unimplemented method 'getProductsByName'");
+    }
```
Comment (Vince-Ku) [inline src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java:41]: Leaving useless code can lead to confusion, as others always have no idea why it was retained.

**Recommendation**
```
// TODO: do something...
```

File: src/main/java/com/orange/main/BaseController.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+/* Transfer Object to ReponseObj, make Object to JSON correctly. */
+    protected static ResponseObj transResponseObj(Object obj){
+        ResponseObj rtnObj = new ResponseObj();
+        rtnObj.setCode("200");
+        rtnObj.setData(obj);
+
+        return rtnObj;
+    }
+
+
```
Comment (Vince-Ku) [inline src/main/java/com/orange/main/BaseController.java]: It seems you lack the error format model, like below

``` JSON
{
   "code":  ...,
   "extra":  ...,  // optional
   "message":  ...,
}
```

File: src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java (hunk 2 above)
Comment (Vince-Ku) [reply src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java:41]: > What if an error occurs in this case?

ditto

**Review by Vince-Ku — CHANGES_REQUESTED** [review]
please check

File: src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java (hunk 1 above)
Comment (asdfg429) [reply src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java:35]: Becaue we didn't discuss about the details of error message. So I prefer to doing this part after we talk about it.

File: src/main/java/com/orange/main/BaseController.java (hunk 3 above)
Comment (asdfg429) [reply src/main/java/com/orange/main/BaseController.java]: Spring Boots responds JSON format is following attributes from Object.
You can see "ReturnObj.java" .

File: src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java (hunk 1 above)
Comment (Vince-Ku) [reply src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java:35]: IMHO, any execution related to the Database might cause exceptions, like, DB connection, data mapping... etc, implementing error handling is nothing to do with displaying error messages (which we could still discuss later) in the frontend, furthermore, you still need to record the log when error occurs, 

maybe we implement error handling at the next commit in this PR?

File: src/main/java/com/orange/main/BaseController.java (hunk 3 above)
Comment (Vince-Ku) [reply src/main/java/com/orange/main/BaseController.java]: I can't get it, would you tell me more details?

File: src/main/java/com/orange/main/products/bo/Products.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+@Entity
+@Table(name = "products")
+public class Products {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String name;
+    private String description;
+    private String types;
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/bo/Products.java]: The type of `types` should be a `enum`.

File: src/main/java/com/orange/main/products/bo/Products.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    private String createdDate;
+    private String updatedDate;
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/bo/Products.java:22]: Try to handle these with JAVA time type. Time type can be calculated and include the timezone.

File: src/main/java/com/orange/main/products/controller/ProductsController.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+@RestController
+@RequestMapping("/v1/products")
+public class ProductsController  extends BaseController{
+    
+    @Autowired
+    private ProductsService productsService;
+
+    @GetMapping("/all")
+    public Object getProducts() {
+
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/controller/ProductsController.java]: style: redundant empty line

File: src/main/java/com/orange/main/products/controller/ProductsController.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        
+        Iterable<Products> product = productsService.getAllProducts();
+        Collection<Products> rt = new ArrayList<>();
+        product.forEach(rt::add);
+        return transResponseObj(rt);
+    }
+
+    @GetMapping("")
+    public Object getProducts(@RequestParam(name = "name",required = false) List<String> name,@RequestParam(name = "sort_by",required = false) String sortBy,@RequestParam(name = "order_by",required = false) String orderBy) {
+        Iterable<Products> product = productsService.findByNameIn(name,sortBy,StringUtils.upperCase(orderBy));
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/controller/ProductsController.java]: ditto
```suggestion
        Iterable<Products> product = productsService.findByNameIn(name, sortBy, StringUtils.upperCase(orderBy));
```

File: src/main/java/com/orange/main/products/controller/ProductsController.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    public Object getProducts() {
+        
+        Iterable<Products> product = productsService.getAllProducts();
+        Collection<Products> rt = new ArrayList<>();
+        product.forEach(rt::add);
+        return transResponseObj(rt);
+    }
+
+    @GetMapping("")
+    public Object getProducts(@RequestParam(name = "name",required = false) List<String> name,@RequestParam(name = "sort_by",required = false) String sortBy,@RequestParam(name = "order_by",required = false) String orderBy) {
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/controller/ProductsController.java]: style: spaces within parameters
```suggestion
    public Object getProducts(@RequestParam(name = "name", required = false) List<String> name, @RequestParam(name = "sort_by", required = false) String sortBy, @RequestParam(name = "order_by", required = false) String orderBy) {
```

Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/controller/ProductsController.java]: The return type being an `object` is too general.

File: src/main/java/com/orange/main/products/bo/Products.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        return this.updatedDate;
+    }
+
+    public void setUpdatedDate(String updatedDate) {
+        this.updatedDate = updatedDate;
+    }
+
+}
+
+
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/bo/Products.java:77]: ditto

File: src/main/java/com/orange/main/products/bo/Products.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    public String getUpdatedDate() {
+        return this.updatedDate;
+    }
+
+    public void setUpdatedDate(String updatedDate) {
+        this.updatedDate = updatedDate;
+    }
+
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/bo/Products.java:71]: ditto

File: src/main/java/com/orange/main/products/controller/ProductsController.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        Iterable<Products> product = productsService.findByNameIn(name,sortBy,StringUtils.upperCase(orderBy));
+        
+        //return new ResponseEntity<>(product, HttpStatus.OK);
+        return transResponseObj(product);
+    }
+
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/controller/ProductsController.java:40]: ditto

File: src/main/java/com/orange/main/products/controller/ProductsController.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    public void setProductsService(ProductsService productsService) {
+        this.productsService = productsService;
+    }
+
+    @GetMapping("/health")
+    public Object getHealth() {
+        
+
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/controller/ProductsController.java]: ditto

File: src/main/java/com/orange/main/products/controller/ProductsController.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    public void setProductsService(ProductsService productsService) {
+        this.productsService = productsService;
+    }
+
+    @GetMapping("/health")
+    public Object getHealth() {
+        
+        
+        return transResponseObj("Egg");
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/controller/ProductsController.java:51]: 🖕

File: src/main/java/com/orange/main/products/repo/ProductsRepository.java [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.data.domain.Sort;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import com.orange.main.products.bo.Products;
+
+public interface ProductsRepository extends JpaRepository<Products, Long>{
+    /* Repository will automatially generate sql according to method name */
+    public Iterable<Products> findByNameIn(List<String> names, Sort sort);//, @Param(value="orderBy") String orderBy
+    //public Iterable<Products> findByNameIn(List<String> name);
+}
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/repo/ProductsRepository.java]: Missing empty line at the end of file.

File: src/main/java/com/orange/main/products/service/ProductsService.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
+package com.orange.main.products.service;
+
+
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/service/ProductsService.java]: ditto

File: src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java:43]: ditto

File: src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    public ProductsRepository getProductsRepository() {
+        return this.productsRepository;
+    }
+
+    public void setProductsRepository(ProductsRepository productsRepository) {
+        this.productsRepository = productsRepository;
+    }
+
+
```
Comment (shadow3x3x3) [inline src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java]: ditto

**Review by shadow3x3x3 — COMMENTED** [review]
Just some comments, but it's not mandatory in this commit.

**Comment by asdfg429** [comment]
@Vince-Ku please review again ,
I'll create another issue for Exception handler

**Review by shadow3x3x3 — APPROVED** [review]

**Pull request merged**